autoagents/agents/agents/search_v3.py

```diff
@@ -0,0 +1,203 @@
+import json
+import uuid
+from datetime import date
+from collections import defaultdict
+from typing import List, Union, Any, Optional, Dict
+
+from langchain.agents import Tool, AgentExecutor, LLMSingleActionAgent, AgentOutputParser
+from langchain.prompts import StringPromptTemplate
+from langchain import LLMChain
+from langchain.schema import AgentAction, AgentFinish
+from langchain.callbacks import get_openai_callback
+from langchain.callbacks.base import AsyncCallbackHandler
+from langchain.callbacks.manager import AsyncCallbackManager
+from langchain.base_language import BaseLanguageModel
+
+from autoagents.agents.tools.tools import search_tool_v3, note_tool_v3, finish_tool_v3
+from autoagents.agents.utils.logger import InteractionsLogger
+
+from pydantic import BaseModel, ValidationError, Extra  # pydantic==1.10.11
+
+
+class InterOutputSchema(BaseModel):
+    thought: str
+    reasoning: str
+    plan: List[str]
+    action: str
+    action_input: str
+    class Config:
+        extra = Extra.forbid
+
+
+class FinalOutputSchema(BaseModel):
+    thought: str
+    reasoning: str
+    plan: List[str]
+    action: str
+    action_input: str
+    citations: List[str]
+    class Config:
+        extra = Extra.forbid
+
+
+def check_valid(o):
+    try:
+        if o.get("action") == "Tool_Finish":
+            FinalOutputSchema(**o)
+        else:
+            InterOutputSchema(**o)
+    except ValidationError:
+        return False
+    return True
+
+
+# Set up a prompt template
+class CustomPromptTemplate(StringPromptTemplate):
+    # The list of tools available
+    tools: List[Tool]
+    ialogger: InteractionsLogger
+
+    def format(self, **kwargs) -> str:
+        # Get the intermediate steps [(AgentAction, Observation)]
+        # Format them in a particular way
+        intermediate_steps = kwargs.pop("intermediate_steps")
+        history = []
+        # Set the agent_scratchpad variable to that value
+        for i, (action, observation) in enumerate(intermediate_steps):
+            if action.tool not in [tool.name for tool in self.tools]:
+                raise Exception("Invalid tool requested by the model.")
+            parsed = json.loads(action.log)
+            if i == len(intermediate_steps) - 1:
+                # Add observation only for the last action
+                parsed["observation"] = observation
+            history.append(parsed)
+        self.ialogger.add_history(history)
+        kwargs["agent_scratchpad"] = json.dumps(history)
```

Comment:
Do we still need to add keys into kwargs?

Reply:
No. Removed. Also added few missing things like tools entry to the log and current date to the goal.